File: frontend/src/app/modules/seguimiento/services/ruta.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable, inject } from '@angular/core';
import * as L from 'leaflet';
import { Observable, catchError, map, of, timeout } from 'rxjs';

import { ApiEnvelope } from '../models/seguimiento.types';

interface RutaData {
  puntos: { latitud: number; longitud: number }[];
}

/**
 * Calcula la ruta por calles reales entre dos puntos (unidad → accidente) vía
 * el proxy de Django hacia OSRM. Si el motor de ruteo falla, no responde a
 * tiempo, o no encuentra ruta, degrada a la línea recta [origen, destino] —
 * esta vista es de monitoreo/visualización, nunca debe romperse ni bloquear
 * por un fallo de un servicio auxiliar de ruteo.
 */
@Injectable({ providedIn: 'root' })
export class RutaService {
  private readonly http = inject(HttpClient);
  private readonly baseUrl = '/api/v1';
  private readonly TIMEOUT_MS = 3000;

  calcularRuta(origen: L.LatLng, destino: L.LatLng): Observable<L.LatLngExpression[]> {
    const recta: L.LatLngExpression[] = [origen, destino];

    return this.http
      .get<ApiEnvelope<RutaData>>(`${this.baseUrl}/seguimiento/ruta`, {
        params: {
          origen: `${origen.lat},${origen.lng}`,
          destino: `${destino.lat},${destino.lng}`,
        },
      })
      .pipe(
        timeout(this.TIMEOUT_MS),
        map((res) =>
          res.data.puntos.length
            ? res.data.puntos.map((p): L.LatLngExpression => L.latLng(p.latitud, p.longitud))
            : recta,
        ),
        catchError(() => of(recta)),
      );
  }
}

File: backend/apps/seguimiento/services/ruta_service.py
"""Servicio de ruteo por calles (OSRM) para el mapa de seguimiento."""

from __future__ import annotations

from core.osrm.client import OsrmClient, OsrmError


class RutaCoordenadasInvalidasError(Exception):
    """Raised when 'origen'/'destino' no son parseables como 'lat,lon'."""


def _parse_coord(raw: str) -> tuple[float, float]:
    try:
        lat_str, lon_str = raw.split(",")
        return float(lat_str), float(lon_str)
    except (ValueError, TypeError) as exc:
        raise RutaCoordenadasInvalidasError("Formato esperado: 'lat,lon'") from exc


class RutaService:
    """
    Proxy delgado hacia OSRM: nunca expone el motor de ruteo directo a
    internet (sin CORS propio, contenedor interno del compose). Si OSRM
    falla, no responde a tiempo, o no encuentra ruta, devuelve `puntos: []`
    — el frontend interpreta una lista vacía como "usar línea recta", ya que
    este servicio es de visualización/monitoreo, no debe bloquear al
    operador por un fallo de un servicio auxiliar.
    """

    def obtener_ruta(self, origen_raw: str, destino_raw: str) -> dict:
        origen_lat, origen_lon = _parse_coord(origen_raw)
        destino_lat, destino_lon = _parse_coord(destino_raw)

        try:
            puntos = OsrmClient().route(origen_lat, origen_lon, destino_lat, destino_lon)
        except OsrmError:
            return {"puntos": []}

        return {"puntos": [{"latitud": lat, "longitud": lon} for lat, lon in puntos]}
